src/Compiler/Opts/ConstantFold.idr

```diff
@@ -0,0 +1,70 @@
+module Compiler.Opts.ConstantFold
+
+import Compiler.CompileExpr
+import Core.Context
+import Core.Context.Log
+import Core.Primitives
+import Core.Value
+import Data.Vect
+
+-- constant folding of primitive functions
+-- if a primitive function is applied to only constant
+-- then replace with the result
+-- if there's only 1 constant argument to a commutative function
+-- move the constant to the right. This simplifies Compiler.Opts.Identity
+constFold : {vars : _} -> CExp vars -> CExp vars
+constFold e@(CLocal _ _) = e
+constFold e@(CRef fc x) = e
+constFold (CLam fc x y) = CLam fc x $ constFold y
+constFold (CLet fc x inlineOK y z) = CLet fc x inlineOK (constFold y) (constFold z)
```

Comment:
Should we inline the `let` if `constFold y` yields a constant?

Reply:
I think that would be a better fit for `Compiler.Inline` (there's some other times where you can inline a let without duplicating work). I also have no idea where the `inlineOK` comes from, so I don'y know if I should change that

Reply:
My point being that `constFold` of `z` where `x` has been replaced by the
constant could fold further.

Reply:
Good point!